### PR TITLE
Update/recipe

### LINKED ISSUE #121: Missing dependencies in Singularity recipe file
Building the Singularity container image does not result in a working container.  The build of Singularity fails due to missing dependencies.

```
$ sudo /usr/local/bin/singularity build sregistry-cli Singularity
... 
+ ./configure --prefix=/usr/local
...
checking for python... no
configure: error: Please install python before installing.
...
checking archive.h usability... no
checking archive.h presence... no
checking for archive.h... no
configure: error: Unable to find the libarchive headers, need package libarchive-devel (libarchive-dev on Debian/Ubuntu)
```

This results in a non-working container:

```
$ ./sregistry-cli pull ...
WARNING Singularity is not installed, function might be limited.
WARNING Singularity is not installed, function might be limited.
...
ERROR Cannot find Singularity! Is it installed?
```

I was able to work around the issue by adding `python` and `libarchive-dev` to the list of packages installed by `apt-get`.

### PR DESCRIPTION
**Pull request by vsoch**
This PR will update both the Singularity and Dockerfile recipes to build the containerized sregistry client. This will close #121. When this is merged, it will be version 0.0.86 on pypi.

### CODE REVIEW

**Comment by samcmill** [comment]
Looks good!

One issue with the Dockerfile recipe.  May need to add `/opt/conda/bin` to `PATH`?

```
...
Building image from sandbox: /tmp/tmp7na6yaaf
Building Singularity image...
Singularity container built: /root/.singularity/shub/nvidia-pytorch:18.03-py3.simg
Cleaning up...
2.5.1-master.gd6e81547
/usr/bin/env: ‘python’: No such file or directory
ERROR Return Code 127: ('/usr/bin/env: ‘python’: No such file or directory\n',)
[container][new] nvidia/pytorch:18.03-py3
Success! /root/.singularity/shub/nvidia-pytorch:18.03-py3.simg
```

**Comment by vsoch** [comment]
Good catch! I'm surprised by this one because the bin of conda should be on the path from `FROM continuumio/miniconda3`. I'll add it to test and let me know if it fixes it.

**Comment by samcmill** [comment]
Hmm... still seeing the same error even after the latest commit.  Something must be clearing the environment?  Tried `--debug`, but no additional messages.

**Comment by samcmill** [comment]
Adding `--no-cache` eliminated the error message.  But I don't understand what is getting cached since it's executing inside a Docker container image?

**Comment by vsoch** [comment]
When you build a Docker image it will default cache the intermediate layers based on the lines in your recipe. If you have the same install command from a remote and the remote changes, it would not see a change in the command and use the cache (old version). This is why with Docker you can develop more quickly, because you don't need to redo work! It's also why for cases like these you need to be careful that the cache isn't tripping you up ;-)

**Comment by samcmill** [comment]
To clarify, `--no-cache` was used as an argument to `sregistry`, not `docker build`.  With the exact same container image, `sregistry pull` exhibits the error, while `sregistry pull --no-cache` does not.

**Comment by vsoch** [comment]
Ah, then it must be that the image existed in your home that is not used when you use the container with no cache, because it does the operation again? I still think this is an issue of having an old version or container or else, but if you want me to debug this I need you to give me the full steps to reproduce the error.

**Comment by vsoch** [comment]
I'm not clear about how you are producing this error - if you are using the Docker image for sregistry you would be issuing commands to it (and not to sregistry on your local machine). Are you shelling into the Docker image first? I need you to show me what exactly isn't working so I can make sure we are good and then merge this PR. Thanks!

**Comment by samcmill** [comment]
```
$ git clone https://github.com/vsoch/sregistry-cli.git
$ cd sregistry-cli
$ git checkout update/recipe
$ docker build -t sregistry-cli -f Dockerfile .
$ export SREGISTRY_NVIDIA_TOKEN="..."
$ docker run --rm -ti --privileged -e SREGISTRY_NVIDIA_TOKEN sregistry-cli pull nvidia://pytorch:18.03-py3
...
Exploding /root/.singularity/docker/metadata/sha256:483f1525239ade069e9ff07fd1111ec4d94b4149cf8e867a50a0fc2e547ea000.tar.gz
Building image from sandbox: /tmp/tmp1tmiag6d
Building Singularity image...
Singularity container built: /root/.singularity/shub/nvidia-pytorch:18.03-py3.simg
Cleaning up...
/usr/bin/env: ‘python’: No such file or directory
/usr/bin/env: ‘python’: No such file or directory
WARNING Inspect command not supported, metadata not included.
[container][new] nvidia/pytorch:18.03-py3
Success! /root/.singularity/shub/nvidia-pytorch:18.03-py3.simg

$ docker run --rm -ti --privileged -e SREGISTRY_NVIDIA_TOKEN sregistry-cli pull --no-cache nvidia://pytorch:18.03-py3
...
Exploding /root/.singularity/docker/metadata/sha256:483f1525239ade069e9ff07fd1111ec4d94b4149cf8e867a50a0fc2e547ea000.tar.gz
Building image from sandbox: /tmp/tmpz18sj42x
Building Singularity image...
Singularity container built: /root/.singularity/shub/nvidia-pytorch:18.03-py3.simg
Cleaning up...
Success! /root/.singularity/shub/nvidia-pytorch:18.03-py3.simg
```

**Comment by vsoch** [comment]
Thank you! I think I see the issue - the `$PATH` isn't being honored, and this could be singularity clearing it or something about using privileged mode (I think likely the first).  Here we can inspect the container (that reproduces the error) to see that the $PATH is correct:

```bash
$docker inspect sregistry-cli
... 
           "Env": [
                "PATH=/opt/conda/bin:/usr/local/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "LANG=C.UTF-8",
                "LC_ALL=C.UTF-8",
                "TINI_VERSION=v0.16.1"
            ],
...
```
`/opt/conda/bin` is not only added, but twice :) So we didn't need to add it again!

And so the issue is adding this line to the file (I'll push this in a second)

```
RUN ln -s /opt/conda/bin/python /usr/bin/python
```

And then the issue is resolved!
```bash
git clone https://github.com/vsoch/sregistry-cli.git
sregistry-cli/
git checkout update/recipe
## NOTE THIS LINE BELOW
docker build --no-cache -t sregistry-cli -f Dockerfile .
docker run --rm -ti --privileged sregistry-cli pull docker://ubuntu
[client|docker] [database|sqlite:////root/.singularity/sregistry.db]
Progress |===================================| 100.0% 
Progress |===================================| 100.0% 
Progress |===================================| 100.0% 
Progress |===================================| 100.0% 
Progress |===================================| 100.0% 
[5/5] |===================================| 100.0% 
Exploding /usr/local/libexec/singularity/bootstrap-scripts/environment.tar
Exploding /root/.singularity/docker/sha256:a48c500ed24e62926cb079df35f964c57d8bb08159b1d29c6a3b0a58dc365dc1.tar.gz
Exploding /root/.singularity/docker/sha256:1e1de00ff7e1fea0858b6a4b5ca208eeca970607ea9a6eb5fc972494e7a0cdde.tar.gz
Exploding /root/.singularity/docker/sha256:0330ca45a200e1fcef05ae97f434366d262a1c50b3dc053e7928b58dd37211dd.tar.gz
Exploding /root/.singularity/docker/sha256:471db38bcfbf0f5bac78012b9d458dfd37309d5cbb99d4e95310321a60a0cfdf.tar.gz
Exploding /root/.singularity/docker/sha256:0b4aba487617ca27519745ae722b8ea1917474c495b91b3c4887728a3f2ee7db.tar.gz
Exploding /root/.singularity/docker/metadata/sha256:c6a9ef4b9995d615851d7786fbc2fe72f72321bee1a87d66919b881a0336525a.tar.gz
Building image from sandbox: /tmp/tmpisys48bd
Building Singularity image...
Singularity container built: /root/.singularity/shub/library-ubuntu:latest.simg
Cleaning up...
{
    "data": {
        "attributes": {
            "labels": null,
            "help": null,
            "deffile": null,
            "runscript": "#!/bin/sh\n\nexec \"/bin/bash\" \"$@\"\n",
            "test": null,
            "environment": "# Custom environment shell code should follow\n\n"
        },
        "type": "container"
    }
}
{
    "data": {
        "attributes": {
            "labels": null,
            "help": null,
            "deffile": null,
            "runscript": "#!/bin/sh\n\nexec \"/bin/bash\" \"$@\"\n",
            "test": null,
            "environment": "# Custom environment shell code should follow\n\n"
        },
        "type": "container"
    }
}
[container][new] library/ubuntu:latest
Success! /root/.singularity/shub/library-ubuntu:latest.simg
```
I'll update the PR in a few minutes (on a different computer)

**Comment by vsoch** [comment]
okay @samcmill please give another go!

File: Dockerfile [hunk 1]
```diff
@@ -36,6 +27,15 @@ RUN git clone https://www.github.com/singularityware/singularity.git
...
+RUN mkdir /code
+ADD . /code
+RUN /opt/conda/bin/pip install setuptools
+RUN /opt/conda/bin/pip install scif
+
+RUN scif install /code/sregistry-cli.scif
+ENTRYPOINT ["sregistry"]
+
 WORKDIR /code
+RUN rm /usr/bin/python && ln -s /opt/conda/bin/python /usr/bin/python
```
Comment (samcmill) [inline Dockerfile]:
```
Step 16/17 : RUN rm /usr/bin/python && ln -s /opt/conda/bin/python /usr/bin/python
 ---> Running in ee1a2bf31b09
rm: cannot remove '/usr/bin/python': No such file or directory
```

Perhaps just `ln -sf /opt/conda/bin/python /usr/bin/python`?

Or `test -f /usr/bin/python || ln -s /opt/conda/bin/python /usr/bin/python`?

**Comment by samcmill** [comment]
After replacing the one Dockerfile line in question with `test -f ...`, everything works.

Looks good!

File: Dockerfile (hunk 1 above)
Comment (vsoch) [reply Dockerfile]: What's the difference between those two?

Comment (vsoch) [reply Dockerfile]: Okay thanks! My preference is for the first, since we want to ensure that the link is correct.

**Comment by vsoch** [comment]
look good now?

**Comment by samcmill** [comment]
+1

**Comment by vsoch** [comment]
Thanks muchly! Uploaded to pypi too!

**Pull request merged**